app/src/main/java/org/mozilla/focus/utils/IntentUtils.kt

```diff
@@ -3,142 +3,68 @@
-     * and true will also be returned). If not handled, we should  fall back to webviews error handling
-     * (which ends up calling our error handling code as needed).
-     *
-     * Note: this method "leaks" the target Uri to Android before asking the user whether they
-     * want to use an external app to open the uri. Ultimately the OS can spy on anything we're
-     * doing in the app, so this isn't an actual "bug".
-     */
-    fun handleExternalUri(context: Context, uri: String): Boolean {
-        // This code is largely based on Fennec's ExternalIntentDuringPrivateBrowsingPromptFragment.java
-        val intent = try { Intent.parseUri(uri, 0) } catch (e: URISyntaxException) { return false }
-
-        // Since we're a browser:
-        intent.addCategory(Intent.CATEGORY_BROWSABLE)
-
-        // This is where we "leak" the uri to the OS. If we're using the system webview, then the OS
-        // already knows that we're opening this uri. Even if we're using GeckoView, the OS can still
-        // see what domains we're visiting, so there's no loss of privacy here:
-        val packageManager = context.packageManager
-        val matchingActivities = packageManager.queryIntentActivities(intent, 0)
-
-        when (matchingActivities.size) {
-            0 -> handleUnsupportedLink(context, intent)
-            1 -> {
-                val info = matchingActivities[0]
-                val externalAppTitle = info?.loadLabel(packageManager) ?: "(null)"
+    fun getIntentChooser(
+        context: Context,
+        intent: Intent,
+        chooserTitle: CharSequence? = null
+    ): Intent? {
+        val chooserIntent: Intent?
+        val resolveInfos = context.packageManager.queryIntentActivities(intent, 0).toHashSet()
+
+        val excludedComponentNames = HashSet<ComponentName>()
+        resolveInfos.forEach {
+            val activityInfo = it.activityInfo
+            if (activityInfo.packageName == context.packageName) {
+                val componentName = ComponentName(activityInfo.packageName, activityInfo.name)
+                excludedComponentNames.add(componentName)
+            }
+        }
 
-                showConfirmationDialog(
-                    context,
-                    intent,
-                    context.getString(R.string.external_app_prompt_title),
-                    R.string.external_app_prompt,
-                    externalAppTitle
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            chooserIntent = Intent.createChooser(intent, chooserTitle)
+                .putExtra(Intent.EXTRA_EXCLUDE_COMPONENTS, excludedComponentNames.toTypedArray())
+        } else {
+            val targetIntents: MutableList<Intent> = ArrayList()
+            for (resolveInfo in resolveInfos) {
+                val activityInfo = resolveInfo.activityInfo
+                if (excludedComponentNames.contains(ComponentName(activityInfo.packageName, activityInfo.name)))
+                    continue
```

Comment:
This suggests we can also work with the result of `resolveInfos.filterNot { .. }`.